benchmarks/Benchmarks.hs

```diff
@@ -25,305 +26,258 @@ import qualified Util.String as US
+--         hmi         = HM.fromList elemsI
+--         hmiSubset   = HM.fromList (takeSubset n elemsI)
+--         hmi2        = HM.fromList elemsI2
+
+--         im          = IM.fromList elemsI
+--         imSubset    = IM.fromList (takeSubset n elemsI)
+
+--         ihm         = IHM.fromList elems
+--         ihmSubset   = IHM.fromList (takeSubset n elems)
+--         ihmbs       = IHM.fromList elemsBS
+--         ihmbsSubset = IHM.fromList (takeSubset n elemsBS)
+--     return Env{..}
+--   where
+--     takeSubset n elements =
+--       -- use 50% of the elements for a subset check.
+--       let subsetSize = round (fromIntegral n * 0.5 :: Double) :: Int
+--       in take subsetSize elements
+
+-- It is important that GHC specializes the benchmarked functions to specific types.
+-- However, this may not happen if the following utility functions are not inlined.
+-- Instead, we force GHC to inline these functions with the rewrite rules below.
+
+type IsKey k = (Ord k, Hashable k, Eq k)
+
+benchMap :: (forall k. IsKey k => Env M.Map k Int -> [Benchmark]) -> Benchmark
+benchMap _ = undefined
+{-# NOINLINE benchMap #-}
+
+benchHashMapPackage :: (forall k. IsKey k => Env IHM.Map k Int -> [Benchmark]) -> Benchmark
+benchHashMapPackage _ = undefined
+{-# NOINLINE benchHashMapPackage #-}
+
+benchIntMap :: (Env IntMap Int Int -> [Benchmark]) -> Benchmark
+benchIntMap _ = undefined
+{-# NOINLINE benchIntMap #-}
+
+benchHashMap :: (forall k. IsKey k => Env HM.HashMap k Int -> [Benchmark]) -> Benchmark
+benchHashMap _ = undefined
+{-# NOINLINE benchHashMap #-}
+
+{-# RULES
+  "force-inline-benchMap" forall (f :: forall k. (IsKey k) => Env M.Map k Int -> [Benchmark]).
+    benchMap f = bgroup "Map" $
+      [ env (setupString M.fromList)     $ \e -> bgroup "String" $ f e
+      , env (setupByteString M.fromList) $ \e -> bgroup "ByteString" $ f e
+      , env (setupInt M.fromList)        $ \e -> bgroup "Int" $ f e
+      ]
+  #-}
+{-# RULES
+  "force-inline-benchHashMapPackage" forall (f :: forall k. (IsKey k) => Env IHM.Map k Int -> [Benchmark]).
+    benchHashMapPackage f = bgroup "hashmap/Map" $
+      [ env (setupString IHM.fromList)     $ \e -> bgroup "String" $ f e
+      , env (setupByteString IHM.fromList) $ \e -> bgroup "ByteString" $ f e
+      , env (setupInt IHM.fromList)        $ \e -> bgroup "Int" $ f e
+      ]
+  #-}
+{-# RULES
+  "force-inline-benchHashMap" forall (f :: forall k. (IsKey k) => Env HM.HashMap k Int -> [Benchmark]).
+    benchHashMap f = bgroup "HashMap" $
+      [ env (setupString HM.fromList)     $ \e -> bgroup "String" $ f e
+      , env (setupByteString HM.fromList) $ \e -> bgroup "ByteString" $ f e
+      , env (setupInt HM.fromList)        $ \e -> bgroup "Int" $ f e
+      ]
+  #-}
+{-# RULES
+  "force-inline-benchIntMap" forall (f :: Env IntMap Int Int -> [Benchmark]).
+    benchIntMap f = env (setupInt (IntMap . IM.fromList)) $ \e -> bgroup "IntMap" $ f e
+  #-}
```

Comment:
I feel somewhat suspicious about adding this kind of complexity to the benchmarks.

Could you say why this is needed, @svenkeidel?

I also wonder whether this PR is the best place to take a deep dive into the benchmarking code. To simplify the review process, I'd slightly prefer keeping this about the original feature request, and to do any larger work on the benchmarks in a separate PR.

Reply:
> I feel somewhat suspicious about adding this kind of complexity to the benchmarks. Could you say why this is needed, @svenkeidel?

As I explained above, the refactoring 1. deduplicates the benchmarks and 2. reduces the number of allocations do make the benchmarks more reproducible. The rewrite rules ensure that benchmarking code is inlined, such that GHC can specialized it. I explain this in this comment: https://github.com/haskell-unordered-containers/unordered-containers/pull/282/commits/57acaec0434b8dffcc8debb9d985cad96e2239a2#diff-e00f41037166e0f4ab0362d1218e6926R179-R181
If we would use INLINE pragmas instead of rewrite rules, GHC may not inline and specialize the benchmark code because the inlined code is too large. Rewrite rules are the best option for forcing the inlining.

> I also wonder whether this PR is the best place to take a deep dive into the benchmarking code. To simplify the review process, I'd slightly prefer keeping this about the original feature request, and to do any larger work on the benchmarks in a separate PR.

I know that this is not the right place to do this. The problem is, as is, the benchmarks are unreliable and not reproducible (different runs produce different results). Therefore, it is hard for me to optimize the performance of `isSubmapOf` without reliable benchmarks.

We can remove the changes to the benchmark suite before we merge this PR. But for now I want to debug the performance with the current benchmark suite.

Reply:
Maybe I put the changes to the benchmark suite in a different branch on top of `subset2`.

Reply:
> If we would use INLINE pragmas instead of rewrite rules, GHC may not inline and specialize the benchmark code because the inlined code is too large. Rewrite rules are the best option for forcing the inlining.

What do you mean with "may not"? Can't you tell from the Core?

I just generally feel that long-running feature branches are a bit of an anti-pattern, since they naturally accumulate merge conflicts and often end up stalling. My impression was that we were largely happy with the library changes, so I'd suggest merging them first, and doing further performance tweaks and/or benchmark refactorings in follow-up PRs.

Reply:
Agreed.